src/coreclr/jit/optimizer.cpp

```diff
@@ -2477,8 +2477,20 @@ PhaseStatus Compiler::optOptimizePostLayout()
         // Reverse conditions to enable fallthrough flow into BBJ_COND's false target
         if (block->KindIs(BBJ_COND) && block->CanRemoveJumpToTarget(block->GetTrueTarget(), this))
         {
-            GenTree* const test = block->lastNode();
+            GenTree* test = block->lastNode();
             assert(test->OperIsConditionalJump());
+
+            if (test->OperIs(GT_JTRUE))
+            {
+                // If we didn't lower a GT_JTRUE node to some conditional IR,
+                // search for the correct node to flip the condition on
+                do
+                {
+                    test = test->gtPrev;
+                    assert(test != nullptr);
+                } while (!test->OperIsCompare() && !test->OperIs(GT_SETCC));
+            }
+
             GenTree* const cond = gtReverseCond(test);
             assert(cond == test); // Ensure `gtReverseCond` did not create a new node
```

Comment:
I don't think this is quite right. There is no invariant on the JTRUE node operand after lowering, so we do need to handle the `GT_NOT` inserted case (or just skip this optimization for these rare cases).

I think the code should rather be something like:
```csharp
if (test->OperIs(GT_JTRUE))
{
  GenTree* const cond = gtReverseCond(test->gtGetOp1());
  if (cond != test->gtGetOp1())
  {
    LIR::AsRange(block).InsertAfter(test->gtGetOp1(), cond);
    test->AsUnOp()->gtOp1 = cond;
  }
}
else
{
  GenTree* const cond = gtReverseCond(test);
  assert(cond == test);
}
```

Skipping the transformation when the terminator node is not JCC, JCMP or JTEST would be fine to me as well.

It might be nice to add a stress mode that `LowerJTrue` uses to leave more JTRUEs in place (`TryLowerConditionToFlagsNode` is almost able to convert all `JTRUE`s today, as you pointed out).

Reply:
I see, thank you for the suggestion! Fixed.

> It might be nice to add a stress mode that LowerJTrue uses to leave more JTRUEs in place (TryLowerConditionToFlagsNode is almost able to convert all JTRUEs today, as you pointed out).

I'll add a stress mode in a follow-up PR. The [lack of diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=704441&view=ms.vss-build-web.run-extensions-tab) seems to further justify exercising these cases.